### ISSUE TITLE
App just crashing when executing simple Entity Framework Core statemente

### ISSUE THREAD

**Issue by MAUIoxo** [xamarin/xamarin-macios#21228]
### Apple platform

iOS

### Framework version

net8.0-*

### Affected platform version

VS 2022 17.11.1, .NET MAUI 8 with 8.0.80/8.0.82, iPhone 14 Pro Max with iOS 17.6.1

### Description

As having talked about all that in the .NET MAUI Discussion with subject [App crash on startup at Entity Framework operations #24556](https://github.com/dotnet/maui/discussions/24556), the App just crashes without useful error indicators in error log.

This crash only occurrs in **`Release`** and was deployed to an iPhone 14 Pro Max with **iOS 17.6.1** and not in `Debug`.
I also tried to debug a `Release` build on iOS Simulator 17.5 which even crashed earlier in the reals App scenario.

Basically, there is just a simple command to get all items from an underlying `SQLite` file database on an iPhone `foodItems = _databaseService.DatabaseContext.Foods.OrderBy(foodItem => foodItem.Name).ToList();`

When this statement is executed the App is just closed and there is no `Exception`

Sentry reports an error of type:
```
EXC_BAD_ACCESS

Exception 1, Code 1, Subcode 12970367616869694461 >
KERN_INVALID_ADDRESS at 0xb40000a0a9bf7bfd.
```

![grafik](https://github.com/user-attachments/assets/6774957f-0532-467b-a0cf-11f8eb8d015a)

We analzed crash reports in the discussion [#24556](https://github.com/dotnet/maui/discussions/24556) but could not find useful hints there.

From the original App, I created a minimal reproducible example. Steps to reproduce, see below

[MacroOptimizerApp.zip](https://github.com/user-attachments/files/16981312/MacroOptimizerApp.zip)

### Steps to Reproduce

1. Build the App `Release`
2. Deploy to iPhone with iOS 17.6.1
3. Start the App
4. The start screen has 2 Tabs: click on the 2nd Tab called "Items"
5. You will see an Alert message that is before the command that leads to the crash
6. Click "OK" > App crashes and is closed

How can you find the relevant part in Code:
1. Open solution
2. Open the OptimizerApp project > ViewModels > OverviewViewModel.cs
3. Method:  `private async void LoadAvailableItemsAsync()`
4. Crash occurs in **line 62**: `foodItems = _databaseService.DatabaseContext.Foods.OrderBy(foodItem => foodItem.Name).ToList();`

### Did you find any workaround?

No workaround

### Relevant logs

```shell
For Error logs please see the original discussion:

https://github.com/dotnet/maui/discussions/24556
```

**Comment by rolfbjarne** [comment]
I get this when I try to build your sample:

```
/Users/rolf/Downloads/MacroOptimizerApp/MacroOptimizerApp/MacroOptimizerApp.csproj : error NU1101: Unable to find package DevExpress.Maui.Core. No packages exist with this id in source(s): /usr/local/share/dotnet/library-packs, configit-sdk, nuget.org
/Users/rolf/Downloads/MacroOptimizerApp/MacroOptimizerApp/MacroOptimizerApp.csproj : error NU1101: Unable to find package DevExpress.Maui.CollectionView. No packages exist with this id in source(s): /usr/local/share/dotnet/library-packs, configit-sdk, nuget.org
```

**Comment by MAUIoxo** [comment]
Thanks for the feedback. In this case I can just provide the following link with instructions:

[DevExpress NuGet Gallery Setup](https://docs.devexpress.com/MAUI/404024/get-started/register-nuget-gallery)

**Comment by MAUIoxo** [comment]
I tried different things now and also removed Sentry from my configuration which did work then and not crash the App. Maybe it was/is Sentry which affects it, maybe not.

I also played with differen other settings in the .NET MAUI Project .csproj and found something that did not crash **including** Sentry.

What really changed it from evil to good was setting the `<RundimeIdentifier>ios-arm64</RuntimeIdentifier>` and creating the `Release|net8.0-ios|AnyCPU` configuration:

```xml
<!-- iOS (Release) -->
<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|net8.0-ios|AnyCPU'">
    <!-- This was almost mandatory -->
    <RuntimeIdentifier>ios-arm64</RuntimeIdentifier>

    <!-- Interpreter disabled since AOT is enabled -->
    <UseInterpreter>false</UseInterpreter>

    <!-- Optimization enabled for Performance -->
    <Optimize>true</Optimize>

    <!-- AOT Compilation enable -->
    <RunAOTCompilation>true</RunAOTCompilation>

    <!-- Link SDK Assemblies Only to Reduce Size -->
    <MtouchLink>SdkOnly</MtouchLink>

    <!-- Use LLVM for Additional Code Optimization -->
    <MtouchUseLlvm>true</MtouchUseLlvm>

    <!-- No Interpretation since AOT is fully used -->
    <MtouchInterpreter>none</MtouchInterpreter>
</PropertyGroup>
```

I commented the settings and what I thought about using them. 

Settings like `<UseInterpreter>true</UseInterpreter>` and `<MtouchInterpreter>all</MtouchInterpreter>` are switched off/removed and also `<RunAOTCompilation>true</RunAOTCompilation>` came back in.

At least with this configuration I could run it on my iPhone so far (just tried with the attached example project).

Using `<UseInterpreter>true</UseInterpreter>` or `<MtouchInterpreter>all</MtouchInterpreter>` has always be a topic of disussion and I am really not able anymore what to say about them when to use and when it is better not to use them.

Do the above settings make sense for a `Release` App or would it be better to change some settings and try? Should the `Android` settings be the same or does it need different attention in some settings?

**Comment by rolfbjarne** [comment]
> I tried different settings again to get to the root cause and also **removed Sentry** from the App to test the impact. When it was removed, it did **not crash** in the attached example! Adding it back in **crashed** it **again**.

This is quite interesting.

My guess is that there's a NullReferenceException somewhere, and Sentry detects that as a crash and terminates the process.

It should be easy enough to verify this theory by triggering a NullReferenceException after initializing Sentry:

```cs
string value = null!;
Console.WriteLine (value.ToString ());
```

and see if that crashes in the same way.

> What I really would name a **Game Changer** was the introduction of the `<RundimeIdentifier>ios-arm64</RuntimeIdentifier>` in a `Release|net8.0-ios|AnyCPU` configuration.

That's somewhat puzzling, `RuntimeIdentifier=ios-arm64` is already the default when building for device from the IDE, or when publishing ("dotnet publish") from the command line (if you're building ("dotnet build") from the command line you need to specify RuntimeIdentifier=ios-arm64, because the default is a simulator build).

> ```
> 
>     
>     <UseInterpreter>false</UseInterpreter>
> ```
> [...]
> ```
> 
>     
>     <MtouchInterpreter>none</MtouchInterpreter>
> ```

This doesn't work as you'd think:

1. `MtouchInterpreter` takes precedence over `UseInterpreter`, so the `UseInterpreter` value is ignored.
2. `MtouchInterpreter=none` means to AOT every assembly, except the assembly named "none".

If you don't want to use the interpreter, just do:

```xml
<UseInterpreter>false</UseInterpreter>
```

> ```
> <RunAOTCompilation>true</RunAOTCompilation>
> ```

This property has no effect at all for iOS projects, you can just remove it.

> What still concerns me when using AOT in the .csproj above is this warning in my .NET Core DLL for the Database.

I believe you might want to enable the interpreter, that should lessen the risk of any problems due to the IL3050 warning. What you can do, is to enable the interpreter, and then AOT-compile every assembly. This still leaves the interpreter present for dynamically generated code, etc.:

```xml
<MtouchInterpreter>-all</MtouchInterpreter>
```

This means: don't interpreter any assembly (but the app still has interpreter support).

Note: you don't need the "UseInterpreter" property in this case.

**Comment by MAUIoxo** [comment]
> ```cs
> string value = null!;
> Console.WriteLine (value.ToString ());
> ```

With a simple provoked crash like this Sentry can detect it was a `NullReferenceException` and it detects it like this and mentions it in its overview. Wrapped around a `try-catch`-Block at least gets us closer to the output we've seen before and which I posted in the issues description above:

![grafik](https://github.com/user-attachments/assets/d47446db-93ce-4170-8c0e-45c040518847)

At least the `AppleCryptoNative_GetRandomBytes` is also available. In the crash report I originally posted, we can also see `sqlite3_rekey_v2` in it, but anyway - very close.

Regarding the introduction `<RundimeIdentifier>ios-arm64</RuntimeIdentifier>` it was like, when I did not use it, also other changes did not noticably provide results. That's why I mentioned this as a _"Game Change"_.

> ```cs
> MtouchInterpreter=none means to AOT every assembly, except the assembly named "none".
> ```

Okay, this would provide the same results then as I wanted. I wanted to try AOT-Compilation for all assemblies. As there was no Assembly named `none` this would have been the same effect. But, I wasn't aware of this meaning, so great you mentioned it to make it clearer!

> If you don't want to use the interpreter, just do:
>
> ```cs
> <UseInterpreter>false</UseInterpreter>
> ```cs

Well, I am not really sure what I want and what the impact is when using the interpreter! I found information on the internet about people that mentioned positive effects and some of them also negative using it.

I also can't really say if using AOT is a burden or a benefit. I think at some time I read that AOT is mandatory on iOS nowadays? But, with how you describe it above this is not true?

Regarding `<MtouchInterpreter>-all</MtouchInterpreter>` what is the difference between using **all** and **-all**? 

<br/>
<br/>

Anyway, I tried it with the following settings:

**.NET MAUI .csproj**:
```xml
<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|net8.0-ios|AnyCPU'">
    <RuntimeIdentifier>ios-arm64</RuntimeIdentifier>
    <Optimize>true</Optimize>
    <AotAssemblies>true</AotAssemblies>
    <MtouchLink>SdkOnly</MtouchLink>
    <MtouchUseLlvm>true</MtouchUseLlvm>
    <MtouchInterpreter>-all</MtouchInterpreter>		
</PropertyGroup>
```

**.NET Core .csproj**:
```xml
<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
    <IsAotCompatible>true</IsAotCompatible>
    <Optimize>true</Optimize>
    <IsTrimmable>False</IsTrimmable>
    <LinkerPleaseIncludeAssemblies>OptimizerDatabase</LinkerPleaseIncludeAssemblies>
</PropertyGroup>
```

<br/>
This again **crashed**!
<br/>
<br/

This one still doesn't even if with your comments does't make fully sense:

```xml
<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|net8.0-ios|AnyCPU'">
	<RuntimeIdentifier>ios-arm64</RuntimeIdentifier>
	<Optimize>true</Optimize>
	<AotAssemblies>false</AotAssemblies>
	<MtouchLink>SdkOnly</MtouchLink>
	<MtouchUseLlvm>true</MtouchUseLlvm>
	<MtouchInterpreter>none</MtouchInterpreter>
</PropertyGroup>
```

If it worked with the above .csproj (the one that is still crashing) would this be considered a recommended setting for such .NET MAUI Apps? Especially asking regarding, **AOT Compilation**, using **Interpreter** and **Optimization**?

<br/>
<br/>

Maybe you still want to give it a try and see if you can get some hints which has the negative impact and leading to this crash?

**Comment by rolfbjarne** [comment]
> I also can't really say if using AOT is a burden or a benefit. I think at some time I read that AOT is mandatory on iOS nowadays? But, with how you describe it above this is not true?

No, AOT is not mandatory.

If AOT or interpreter is best depends on the app: interpreted code is typically smaller (resulting in smaller apps), while AOT compiled code is typically faster (but it might not even be measurable). Unless either size or performance show up as a problem during your own testing, then I wouldn't worry about the specific interpreter/AOT configuration.

> Regarding `<MtouchInterpreter>-all</MtouchInterpreter>` what is the difference between using **all** and **-all**?

* "all": interpret all assemblies
* "-all": do not interpret any assemblies

> Maybe you still want to give it a try and see if you can get some hints which has the negative impact and leading to this crash?

Are you able to reproduce a crash without Sentry?

**Comment by MAUIoxo** [comment]
Thanks for all the information you shared - this helped a lot!

**Without Sentry**:
I checked without Sentry and different Build Settings. So far, it looks like I could not get any setting that crashed without Sentry! I also tried removing `<RuntimeIdentifier>ios-arm64</RuntimeIdentifier>`. This wasn't possible with all settings as described above and also crashed sometimes. But, I concentrated more on the other settings.

**Including Sentry**:
The Build Settings I shared above and which did not crash including Sentry were strange in a way. I had `<MtouchInterpreter>none</MtouchInterpreter>` which seems to work and this would be _"Activate Interpreter, but don't use it actually"_, (for my assemblies as there is no Assembly with name "none") if I understood it correctly. But, `<MtouchInterpreter>-all</MtouchInterpreter>` acutally crashed. Strange in a way

**Comment by MAUIoxo** [comment]
This is really interesting!

I was deploying it really to my own physical device. iPhone 14 Pro Max which currently has iOS 17.6.1

**Comment by MAUIoxo** [comment]
Has anything fundamental changed with the recent Update to **Visual Studio 17.11.4**?
I updated yesterday and deployed the original App again without changes and it did not crash on the iPhone...

**Comment by rolfbjarne** [comment]
> Has anything fundamental changed with the recent Update to **Visual Studio 17.11.4**? I updated yesterday and deployed the original App again without changes and it did not crash on the iPhone...

No, nothing's changed... so maybe the previous update/installation had a weird bug for some reason?

Since the problem is gone, I suppose we can close this then?

**Comment by MAUIoxo** [comment]
Yes, we can close both issues (also https://github.com/dotnet/maui/discussions/24556) I would say.

Don't know if also workloads got updated or whatever it solved it...

Just wanted to say thank you for all your support on that. I really appreciate that! Also for all the details you shared - it helped a lot
